sql: hide auto-created rowid column

madelynnblue opened the pull request:
Mark the rowid column (auto created for tables without primary keys)
as hidden, which is a new field of ColumnDescriptor.

Do not use any hidden columns in implied column lists (like SELECT *
or INSERT with no column listing). The rowid column still shows up
in places like SHOW COLUMNS, and can be manually specified in any
operation (SELECT rowid from t).

Fixes #3293

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3318)

<!-- Reviewable:end -->

tamird commented:
Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[sql/insert.go, line 246 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3318#-K4dnKOzA-IXG4VL3vx8:-K4dnKOzA-IXG4VL3vx9:1366473640) ([raw file](https://github.com/cockroachdb/cockroach/blob/d17d81bd11b6b7ddffa6ea8213cb23f90bb0b1ff/sql/insert.go#L246)):*
I think this logic should move into an accessor on TableDescriptor, but that might be uglier in practice.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3318)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 5 of 7 files reviewed at latest revision, 1 unresolved discussion.

---

*[sql/insert.go, line 246 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3318#-K4dnKOzA-IXG4VL3vx8:-K4dpaJ65CqE9wb6BoVr:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d17d81bd11b6b7ddffa6ea8213cb23f90bb0b1ff/sql/insert.go#L246)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3318)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r2.
Review status: 6 of 7 files reviewed at latest revision, 1 unresolved discussion.

---

*[sql/insert.go, line 246 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3318#-K4dnKOzA-IXG4VL3vx8:-K4dpmlRVYqYfjEG3XZP:-11541948) ([raw file](https://github.com/cockroachdb/cockroach/blob/d17d81bd11b6b7ddffa6ea8213cb23f90bb0b1ff/sql/insert.go#L246)):*
Can you also use this accessor in the other places that care about visibility? My comment should really have been "inspecting column.Hidden should only happen in the accessor".

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3318)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 5 of 7 files reviewed at latest revision, 1 unresolved discussion.

---

*[sql/insert.go, line 246 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3318#-K4dnKOzA-IXG4VL3vx8:-K4dqn365DkddOB9utsH:1178432605) ([raw file](https://github.com/cockroachdb/cockroach/blob/d17d81bd11b6b7ddffa6ea8213cb23f90bb0b1ff/sql/insert.go#L246)):*
I've replaced a second instance of inspecting col.Hidden with using VisibleColumns. Now it's down to just one place. That place is where it iterates over the column IDs in the index and then looks up the column. No obvious API comes to mind that is better than the current PR.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3318)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
sql: hide auto-created rowid column

Mark the rowid column (auto created for tables without primary keys)
as hidden, which is a new field of ColumnDescriptor.

Do not use any hidden columns in implied column lists (like SELECT *
or INSERT with no column listing). The rowid column still shows up
in places like SHOW COLUMNS, and can be manually specified in any
operation (SELECT rowid from t).

Fixes #3293

tamird commented:
Reviewed 2 of 2 files at r3.
Review status: 6 of 7 files reviewed at latest revision, 1 unresolved discussion.

---

*[sql/structured.go, line 556 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3318#-K4dsWgSXFC2Vm8gwcqe:-K4dsWgSXFC2Vm8gwcqf:247262164) ([raw file](https://github.com/cockroachdb/cockroach/blob/62ccded0e0b67cd3818830b0f6721f2de3f210dd/sql/structured.go#L556)):*
this doesn't need to be exported.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3318)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/structured.go, line 557 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3318#-K4dtanvRbGcc0fX2bm_:-K4dtanvRbGcc0fX2bma:1171970253) ([raw file](https://github.com/cockroachdb/cockroach/blob/c20df95a85235221a0b88ab37785820be790989c/sql/structured.go#L557)):*
consider pre-allocating this to `len(desc.Columns)`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3318)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/structured.go, line 557 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3318#-K4dtanvRbGcc0fX2bm_:-K4edfvcc2dUN5z_1Fqo:1899726837) ([raw file](https://github.com/cockroachdb/cockroach/blob/c20df95a85235221a0b88ab37785820be790989c/sql/structured.go#L557)):*
Alternately, consider sorting hidden columns after visible columns so that you can return a prefix of `desc.Columns`. 

---

*[sql/testdata/no_primary_key, line 66 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3318#-K4ee9z-Zr7NfB6pOZ-t:-K4ee9z-Zr7NfB6pOZ-u:-334474761) ([raw file](https://github.com/cockroachdb/cockroach/blob/c20df95a85235221a0b88ab37785820be790989c/sql/testdata/no_primary_key#L66)):*
Add a test case here that adds an index to the table. Verify that you can query the index explicitly (via the `table@index` syntax).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3318)*
<!-- Sent from Reviewable.io -->

mjibson commented:
I'm abandoning this PR because it's causing weird errors with uniqueness of index errors. It is clear that the number of one-offs required to make non-PK tables work based on this current implementation is not acceptable, and will likely continue to grow and add lots of brittleness to the system.

petermattis commented:
Can you describe in more detail the problem with unique indexes and this approach?

mjibson commented:
I ran this test:

```
debian:26257> CREATE TABLE t (a int);
OK
debian:26257> INSERT INTO t VALUES (1);
OK
debian:26257> INSERT INTO t VALUES (1);
OK
debian:26257> CREATE INDEX a_idx ON t (a);
query error: duplicate key value (a)=(1) violates unique constraint "a_idx"
debian:26257> CREATE TABLE v (a int primary key, b int);
OK
debian:26257> INSERT INTO v VALUES (1, 1);
OK
debian:26257> INSERT INTO v VALUES (2, 1);
OK
debian:26257> CREATE INDEX b_idx ON v (b);
OK
```

What that does is first create a table with no primary key and insert two identical rows. Creating an index on the only (user-visible) column on that table fails with a uniqueness failure, which is unexpected since I didn't specify a unique table.

Then, do exactly the same test but with a table that specifies a primary key, and it works as expected.

There's a problem somewhere with the no-PK table that is causing this.

petermattis commented:
Got it. Have you identified what the cause of the problem was? Should be straightforward to track down where/why that error originated. I'm trying to understand where the brittleness is coming from and wondering if it will affect any solution to non-PK tables.

mjibson commented:
Not sure about the cause. I stopped pursuing this after I found the above error.
